man/pry.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "PRY" "1" "December 2012" "" ""
.
.SH "NAME"
\fBPRY\fR \- A Reference to the PRY repl\.
.
.SH "Synopsis"
\fBpry\fR [\fB\-\-version\fR] [\fB\-\-exec\fR] [\fB\-\-no\-pager\fR] [\fB\-\-no\-history\fR] [\fB\-\-no\-color\fR] [\fB\-f\fR] [\fB\-\-no\-plugins\fR] [\fB\-\-installed\-plugins\fR] [\fB\-\-simple\-prompt\fR] [\fB\-\-require\fR \fIfile\fR] [\fB\-I\fR] [\fB\-\-context\fR] [\fB\-\-help\fR]
.
.SH "DESCRIPTION"
Pry is a powerful alternative to the standard IRB shell for Ruby\. It is written from scratch to provide a number of advanced features\.
.
.SH "HOMEPAGE"
http://pry\.github\.com/
.
.SH "OPTIONS"
.
.TP
\fB\-v \-\-version\fR
Prints the version of Pry\.
.
.TP
\fB\-e \-\-exec\fR
Executes argument in context before the session starts\.
.
.TP
\fB\-\-no\-pager\fR
Disable pager for long output\.
.
.TP
\fB\-\-no\-history\fR
Disable history loading\.
.
.TP
\fB\-\-no\-color\fR
Disable syntax highlighting for session\.
.
.TP
\fB\-f\fR
Prevent loading of ~/\.pryrc for session\.
.
.TP
\fB\-\-no\-plugins\fR
Supress loading of plugins\.
.
.TP
\fB\-\-installed\-plugins\fR
List installed plugins\.
.
.TP
\fB\-\-simple\-prompt\fR
Enable simple prompt mode (eg, >>)\.
.
.TP
\fB\-r \-\-require\fR
Require a ruby script at startup\.
.
.TP
\fB\-I\fR
Add a path to the $LOAD_PATH
.
.TP
\fB\-c \-\-context\fR
Start the session in the specified context\. Equivalent to \fBcontext\.pry\fR in a session\.
.
.SH "FILES"
~/\.pryrc Personal pry initialization
.
.SH "Pry Commands"
.
.TP
\fBtoggle\-color\fR
Toggle syntax highlighting\. Toggle syntax highlighting\.
.
.TP
\fBsimple\-prompt\fR
Toggle the simple prompt\. Toggle the simple prompt\.
.
.TP
\fBpry\-version\fR
Show Pry version\. Show Pry version\.
.
.TP
\fBreload\-method\fR
Reload the source file that contains the specified method Reload the source file that contains the specified method
.
.TP
\fBhelp\fR
Show a list of commands\. Type \fBhelp <foo>\fR for information about \fIfoo\fR\.
.
.IP "" 4
.
.nf

      Usage: help [ COMMAND ]

    With no arguments, help lists all the available commands in the current
    command\-set along with their description\.

    When given a command name as an argument, shows the help for that command\.
.
.fi
.
.IP "" 0

.
.TP
\fBgem\-install\fR
Install a gem and refresh the gem cache\.
.
.IP "" 4
.
.nf

      Usage: gem\-install GEM_NAME

    Installs the given gem and refreshes the gem cache so that you can immediately \'require GEM_FILE\'
.
.fi
.
.IP "" 0

.
.TP
\fBgem\-cd\fR
Change working directory to specified gem\'s directory\.
.
.IP "" 4
.
.nf

      Usage: gem\-cd GEM_NAME

    Change the current working directory to that in which the given gem is installed\.
.
.fi
.
.IP "" 0

.
.TP
\fBgem\-list\fR
List and search installed gems\.
.
.IP "" 4
.
.nf

      Usage: gem\-list [REGEX]

    List all installed gems, when a regex is provided, limit the output to those that
    match the regex\.
.
.fi
.
.IP "" 0

.
.TP
\fBls\fR
Show the list of vars and methods in the current scope\. Show the list of vars and methods in the current scope\.
.
.TP
\fBcd\fR
Move into a new context (object or scope)\.
.
.IP "" 4
.
.nf

      Usage: cd [OPTIONS] [\-\-help]

    Move into new context (object or scope)\. As in unix shells use
    `cd \.\.` to go back, `cd /` to return to Pry top\-level and `cd \-`
    to toggle between last two scopes)\.
    Complex syntax (e\.g `cd \.\./@x/y`) also supported\.

    e\.g: `cd @x`
    e\.g: `cd \.\.`
    e\.g: `cd /`
    e\.g: `cd \-`

    https://github\.com/pry/pry/wiki/State\-navigation#wiki\-Changing_scope
.
.fi
.
.IP "" 0

.
.TP
\fBfind\-method\fR
Recursively search for a method within a Class/Module or the current namespace\. find\-method [\-n | \-c] METHOD [NAMESPACE]
.
.IP "" 4
.
.nf

      Usage: find\-method  [\-n | \-c] METHOD [NAMESPACE]

    Recursively search for a method within a Class/Module or the current namespace\.
    Use the `\-n` switch (the default) to search for methods whose name matches the given regex\.
    Use the `\-c` switch to search for methods that contain the given code\.

    e\.g find\-method re Pry                # find all methods whose name match /re/ inside the Pry namespace\. Matches Pry#repl, etc\.
    e\.g find\-method \-c \'output\.puts\' Pry  # find all methods that contain the code: output\.puts inside the Pry namepsace\.
.
.fi
.
.IP "" 0

.
.TP
\fBwhereami\fR
Show code surrounding the current context\.
.
.IP "" 4
.
.nf

      Usage: whereami [\-q] [N]

    Describe the current location\. If you use `binding\.pry` inside a
    method then whereami will print out the source for that method\.

    If a number is passed, then N lines before and after the current line
    will be shown instead of the method itself\.

    The `\-q` flag can be used to suppress error messages in the case that
    there\'s no code to show\. This is used by pry in the default
    before_session hook to show you when you arrive at a `binding\.pry`\.

    When pry was started on an Object and there is no associated method,
    whereami will instead output a brief description of the current
    object\.
.
.fi
.
.IP "" 0

.
.TP
\fBpry\-backtrace\fR
Show the backtrace for the Pry session\.
.
.IP "" 4
.
.nf

      Usage:   pry\-backtrace [OPTIONS] [\-\-help]

    Show the backtrace for the position in the code where Pry was started\. This can be used to
    infer the behavior of the program immediately before it entered Pry, just like the backtrace
    property of an exception\.

    (NOTE: if you are looking for the backtrace of the most recent exception raised,
    just type: `_ex_\.backtrace` instead, see https://github\.com/pry/pry/wiki/Special\-Locals)

    e\.g: pry\-backtrace
.
.fi
.
.IP "" 0

.
.TP
\fBreset\fR
Reset the REPL to a clean state\. Reset the REPL to a clean state\.
.
.TP
\fB(?\-mix:wtf([?!]*))\fR
Show the backtrace of the most recent exception
.
.IP "" 4
.
.nf

      Show\'s a few lines of the backtrace of the most recent exception (also available
    as _ex_\.backtrace)\.

    If you want to see more lines, add more question marks or exclamation marks:

    e\.g\.
    pry(main)> wtf?
    pry(main)> wtf?!???!?!?

    To see the entire backtrace, pass the \-v/\-\-verbose flag:

    e\.g\.
    pry(main)> wtf \-v
.
.fi
.
.IP "" 0

.
.TP
\fB(?\-mix:raise\-up(!?\eb\.*))\fR
Raise an exception out of the current pry instance\.
.
.IP "" 4
.
.nf

      Raise up, like exit, allows you to quit pry\. Instead of returning a value however, it raises an exception\.
    If you don\'t provide the exception to be raised, it will use the most recent exception (in pry _ex_)\.

    e\.g\. `raise\-up "get\-me\-out\-of\-here"` is equivalent to:
         `raise "get\-me\-out\-of\-here"
          raise\-up`

    When called as raise\-up! (with an exclamation mark), this command raises the exception through
    any nested prys you have created by "cd"ing into objects\.
.
.fi
.
.IP "" 0

.
.TP
\fBswitch\-to\fR
Start a new sub\-session on a binding in the current stack (numbered by nesting)\. Start a new sub\-session on a binding in the current stack (numbered by nesting)\.
.
.TP
\fBnesting\fR
Show nesting information\. Show nesting information\.
.
.TP
\fBjump\-to\fR
Jump to a binding further up the stack, popping all bindings below\. Jump to a binding further up the stack, popping all bindings below\.
.
.TP
\fBexit\-all\fR
End the current Pry session (popping all bindings) and returning to caller\. Accepts optional return value\. Aliases: !!@ End the current Pry session (popping all bindings) and returning to caller\. Accepts optional return value\. Aliases: !!@
.
.TP
\fB!!@\fR
Alias for \fBexit\-all\fR Alias for \fBexit\-all\fR
.
.TP
\fBexit\fR
Pop the previous binding (does NOT exit program)\. Aliases: quit
.
.IP "" 4
.
.nf

      Usage:   exit [OPTIONS] [\-\-help]
    Aliases: quit

    It can be useful to exit a context with a user\-provided value\. For
    instance an exit value can be used to determine program flow\.

    e\.g: `exit "pry this"`
    e\.g: `exit`

    https://github\.com/pry/pry/wiki/State\-navigation#wiki\-Exit_with_value
.
.fi
.
.IP "" 0

.
.TP
\fBquit\fR
Alias for \fBexit\fR Alias for \fBexit\fR
.
.TP
\fBexit\-program\fR
End the current program\. Aliases: quit\-program, !!! End the current program\. Aliases: quit\-program, !!!
.
.TP
\fBquit\-program\fR
Alias for \fBexit\-program\fR Alias for \fBexit\-program\fR
.
.TP
\fB!!!\fR
Alias for \fBexit\-program\fR Alias for \fBexit\-program\fR
.
.TP
\fB!pry\fR
Start a Pry session on current self; this even works mid multi\-line expression\. Start a Pry session on current self; this even works mid multi\-line expression\.
.
.TP
\fBhist\fR
Show and replay Readline history\. Aliases: history
.
.IP "" 4
.
.nf

      Usage: hist
           hist \-\-head N
           hist \-\-tail N
           hist \-\-show START\.\.END
           hist \-\-grep PATTERN
           hist \-\-clear
           hist \-\-replay START\.\.END
           hist \-\-save [START\.\.END] FILE
.
.fi
.
.IP "" 0

.
.TP
\fBhistory\fR
Alias for \fBhist\fR Alias for \fBhist\fR
.
.TP
\fB!\fR
Clear the input buffer\. Useful if the parsing process goes wrong and you get stuck in the read loop\. Clear the input buffer\. Useful if the parsing process goes wrong and you get stuck in the read loop\.
.
.TP
\fBshow\-input\fR
Show the contents of the input buffer for the current multi\-line expression\. Show the contents of the input buffer for the current multi\-line expression\.
.
.TP
\fBedit\fR
Invoke the default editor on a file\.
.
.IP "" 4
.
.nf

      Usage: edit [\-\-no\-reload|\-\-reload] [\-\-line LINE] [\-\-temp|\-\-ex|FILE[:LINE]|\-\-in N]

    Open a text editor\. When no FILE is given, edits the pry input buffer\.
    Ensure Pry\.config\.editor is set to your editor of choice\.

    e\.g: `edit sample\.rb`
    e\.g: `edit sample\.rb \-\-line 105`
    e\.g: `edit \-\-ex`

    https://github\.com/pry/pry/wiki/Editor\-integration#wiki\-Edit_command
.
.fi
.
.IP "" 0

.
.TP
\fBedit\-method\fR
Edit the source code for a method\.
.
.IP "" 4
.
.nf

      Usage: edit\-method [OPTIONS] [METH]

    Edit the method METH in an editor\.
    Ensure Pry\.config\.editor is set to your editor of choice\.

    e\.g: `edit\-method hello_method`
    e\.g: `edit\-method Pry#rep`
    e\.g: `edit\-method`

    https://github\.com/pry/pry/wiki/Editor\-integration#wiki\-Edit_method
.
.fi
.
.IP "" 0

.
.TP
\fB(?\-mix:amend\-line(?: (\-?\ed+)(?:\e\.\e\.(\-?\ed+))?)?)\fR
Amend a line of input in multi\-line mode\.
.
.IP "" 4
.
.nf

      Amend a line of input in multi\-line mode\. `amend\-line N`, where the N in `amend\-line N` represents line to replace\.

    Can also specify a range of lines using `amend\-line N\.\.M` syntax\. Passing \'!\' as replacement content deletes the line(s) instead\.
    e\.g amend\-line 1 puts \'hello world! # replace line 1\'
    e\.g amend\-line 1\.\.4 !               # delete lines 1\.\.4
    e\.g amend\-line 3 >puts \'goodbye\'    # insert before line 3
    e\.g amend\-line puts \'hello again\'   # no line number modifies immediately preceding line
.
.fi
.
.IP "" 0

.
.TP
\fBplay\fR
Play back a string variable or a method or a file as input\.
.
.IP "" 4
.
.nf

      Usage: play [OPTIONS] [\-\-help]

    The play command enables you to replay code from files and methods as
    if they were entered directly in the Pry REPL\. Default action (no
    options) is to play the provided string variable

    e\.g: `play \-i 20 \-\-lines 1\.\.3`
    e\.g: `play \-m Pry#repl \-\-lines 1\.\.\-1`
    e\.g: `play \-f Rakefile \-\-lines 5`

    https://github\.com/pry/pry/wiki/User\-Input#wiki\-Play
.
.fi
.
.IP "" 0

.
.TP
\fBgist\fR
Gist a method or expression history to GitHub\.
.
.IP "" 4
.
.nf

      Usage: gist [OPTIONS] [METH]
    Gist method (doc or source) or input expression to GitHub\.

    If you\'d like to permanently associate your gists with your GitHub account run `gist \-\-login`\.

    e\.g: gist \-m my_method       # gist the method my_method
    e\.g: gist \-d my_method       # gist the documentation for my_method
    e\.g: gist \-i 1\.\.10           # gist the input expressions from 1 to 10
    e\.g: gist \-k show\-method     # gist the command show\-method
    e\.g: gist \-c Pry             # gist the Pry class
    e\.g: gist \-m hello_world \-\-lines 2\.\.\-2    # gist from lines 2 to the second\-last of the hello_world method
    e\.g: gist \-m my_method \-\-clip             # Copy my_method source to clipboard, do not gist it\.
.
.fi
.
.IP "" 0

.
.TP
\fBclipit\fR
Alias for \fBgist \-\-clip\fR Alias for \fBgist \-\-clip\fR
.
.TP
\fBjist\fR
Alias for \fBgist\fR Alias for \fBgist\fR
.
.TP
\fB(?\-mix:\e\.(\.*))\fR
All text following a \'\.\' is forwarded to the shell\. All text following a \'\.\' is forwarded to the shell\.
.
.TP
\fBshell\-mode\fR
Toggle shell mode\. Bring in pwd prompt and file completion\. Toggle shell mode\. Bring in pwd prompt and file completion\.
.
.TP
\fBfile\-mode\fR
Alias for \fBshell\-mode\fR Alias for \fBshell\-mode\fR
.
.TP
\fBsave\-file\fR
Export to a file using content from the REPL\.
.
.IP "" 4
.
.nf

      Usage: save\-file [OPTIONS] [FILE]
    Save REPL content to a file\.
    e\.g: save\-file \-m my_method \-m my_method2 \./hello\.rb
    e\.g: save\-file \-i 1\.\.10 \./hello\.rb \-\-append
    e\.g: save\-file \-k show\-method \./my_command\.rb
    e\.g: save\-file \-f sample_file \-\-lines 2\.\.10 \./output_file\.rb
.
.fi
.
.IP "" 0

.
.TP
\fBcat\fR
Show code from a file, Pry\'s input buffer, or the last exception\.
.
.IP "" 4
.
.nf

      Usage: cat FILE
           cat \-\-ex [STACK_INDEX]
           cat \-\-in [INPUT_INDEX_OR_RANGE]

    cat is capable of showing part or all of a source file, the context of the
    last exception, or an expression from Pry\'s input history\.

    cat \-\-ex defaults to showing the lines surrounding the location of the last
    exception\. Invoking it more than once travels up the exception\'s backtrace,
    and providing a number shows the context of the given index of the backtrace\.
.
.fi
.
.IP "" 0

.
.TP
\fBshow\-doc\fR
Show the documentation for a method or class\. Aliases: ?
.
.IP "" 4
.
.nf

      Usage: show\-doc [OPTIONS] [METH]
    Aliases: ?

    Show the documentation for a method or class\. Tries instance methods first and then methods by default\.
    e\.g show\-doc hello_method    # docs for hello_method
    e\.g show\-doc Pry             # docs for Pry class
    e\.g show\-doc Pry \-a          # docs for all definitions of Pry class (all monkey patches)
.
.fi
.
.IP "" 0

.
.TP
\fB?\fR
Alias for \fBshow\-doc\fR Alias for \fBshow\-doc\fR
.
.TP
\fBstat\fR
View method information and set \fIfile\fR and \fIdir\fR locals\.
.
.IP "" 4
.
.nf

        Usage: stat [OPTIONS] [METH]
      Show method information for method METH and set _file_ and _dir_ locals\.
      e\.g: stat hello_method
.
.fi
.
.IP "" 0

.
.TP
\fBshow\-source\fR
Show the source for a method or class\. Aliases: $, show\-method
.
.IP "" 4
.
.nf

      Usage: show\-source [OPTIONS] [METH|CLASS]
    Aliases: $, show\-method

    Show the source for a method or class\. Tries instance methods first and then methods by default\.

    e\.g: `show\-source hello_method`
    e\.g: `show\-source \-m hello_method`
    e\.g: `show\-source Pry#rep`         # source for Pry#rep method
    e\.g: `show\-source Pry`             # source for Pry class
    e\.g: `show\-source Pry \-a`          # source for all Pry class definitions (all monkey patches)

    https://github\.com/pry/pry/wiki/Source\-browsing#wiki\-Show_method
.
.fi
.
.IP "" 0

.
.TP
\fBshow\-method\fR
Alias for \fBshow\-source\fR Alias for \fBshow\-source\fR
.
.TP
\fB$\fR
Alias for \fBshow\-source\fR Alias for \fBshow\-source\fR
.
.TP
\fBshow\-command\fR
Show the source for CMD\. Show the source for CMD\.
.
.TP
\fBri\fR
View ri documentation\. e\.g \fBri Array#each\fR
.
.IP "" 4
.
.nf

      Usage: ri [spec]
    e\.g\. ri Array#each

    Relies on the rdoc gem being installed\. See also: show\-doc\.
.
.fi
.
.IP "" 0

.
.TP
\fBnyan\-cat\fR:

.
.TP
\fB(?\-mix:!s\e/(\.*?)\e/(\.*?))\fR:

.
.TP
\fBget\-naked\fR:

.
.TP
\fBeast\-coker\fR:

.
.TP
\fBcohen\-poem\fR:

.
.TP
\fBtest\-ansi\fR:

.
.TP
\fBimport\-set\fR
Import a command set Import a command set
.
.TP
\fBinstall\-command\fR
Install a disabled command\.
.
.IP "" 4
.
.nf

      Usage: install\-command COMMAND

    Installs the gems necessary to run the given COMMAND\. You will generally not
    need to run this unless told to by an error message\.
.
.fi
.
.IP "" 0

.
.TP
\fBcopy\fR
Copy argument to the clip\-board Copy argument to the clip\-board
.
.TP
\fBclear\fR
No description\. No description\.
.
.TP
\fBsql\fR
Send sql over AR\. Send sql over AR\.
.
.TP
\fBcaller_method\fR
No description\. No description\.
.
.TP
\fBgrowl\fR
No description\. No description\.
.
.SH "EXAMPLES"
.
.SS "Basic Usage"
.
.nf

$ pry
[1] pry(main)>4 + 5
=> 9
[2] pry(main)> def hello_world
[2] pry(main)*   puts "Hello, World!"
[2] pry(main)* end
=> nil
[3] pry(main)> hello_world
Hello, World!
=> nil
.
.fi
.
.SS "Command Line Interaction"
Prefix any command you want your shell to execute with a period and pry will return the results from your shell\.
.
.IP "" 4
.
.nf

[1] pry(main)> \.date
Fri Nov 11 09:52:07 EST 2011
.
.fi
.
.IP "" 0
.
.P
On the command line enter \fBshell\-mode\fR to incorporate the current working directory into the Pry prompt\.
.
.IP "" 4
.
.nf

pry(main)> shell\-mode
pry main:/Users/john/ruby/projects/pry $ \.cd \.\.
pry main:/Users/john/ruby/projects $ \.cd ~
pry main:/Users/john $ \.pwd
/Users/john
pry main:/Users/john $ shell\-mode
pry(main)>
.
.fi
.
.IP "" 0
.
.SS "State Navigation"
The cd command is used to move into a new object (or scope) inside a Pry session\. When inside the new scope it becomes the self for the session and all commands and methods will operate on this new self\.
.
.IP "" 4
.
.nf

pry(main)> self
=> main
pry(main)> cd Pry
pry(Pry):1> self
=> Pry
pry(Pry):1> cd \.\.
pry(main)>
.
.fi
.
.IP "" 0
.
.P
The ls command is essentially a unified wrapper to a number of Ruby\'s introspection mechanisms, including (but not limited to) the following methods: methods, instance_variables, constants, local_variables, instance_methods, class_variables and all the various permutations thereof\.
.
.P

CHANGELOG
14/07/2012 version 0.9.10
dependency changes:
* #561 upgrade the slop gem to version 3
* #590 move to the jist gem from gist.
* upgrade method_source to 0.8

new features:
* #572 add --hist, -o and -k flags to gist command
* #584 support show-source/doc on methods defined in class-eval
* #585 support show-source/doc on gem methods defined in C
* #596 add --disable-plugin and --select-plugin options
* #597 allow "cd -" to switch between bindings
* #612 add Pry.config.should_load_local_rc to turn off ./.pryrc
* allow running a file of pry input with pry <file>
* support colours in "ri" command
* add before_eval hook
* prompt now gets a lot more data when proc arity is 1

bug fixes &c.
* #554 removed the "req" command
* #567 fix rendering bugs when starting pry
* #568 fix Array#pretty_print on Jruby
* #575 fix "edit" on windows
* #576 fix "find-method" in the presence of badly behaved objects
* #580 fix "whereami" in erb files on rails
* #632 raise fewer exceptions while tab completing
* #605 dont immediately quite pry when an error happens in readline
* #606 support for ansicon to give jruby windows users colour
* #613 massive speed improvements to show-source for modules
* #620 improve whereami command when not in a binding.pry
* #622 support embedded documents (=begin ... =end)
* #627 support editing files with spaces in the name
* changed __binding_impl__ to __pry__
* support for absolute paths in $EDITOR
* fix "cat" command on files with unknown extensions
* many many internal refactorings and tidyings

09/05/2012 version 0.9.9.6 fixes #558
* #558 has been a thorn in our side for long enough, hopefully this properly fixes it
(though im not to confident :P)

09/05/2012 version 0.9.9.5 minor bugfix
* fixed ZeroDivisionError in correct_indentation, bug #558
* fix double highlighting in rdoc, bug #562
* autocreate configuration for plugins, bug #548

26/4/2012 version 0.9.9.4 major bugfix
* fixed `NoMethodError: undefined method `winsize' for #<IO:<STDOUT>>`, bug #549
* fixes for jruby
* breakage on `exit` syntax error, fixes, #550
* heredoc content no longer auto-indented

19/4/2012 version 0.9.9.3 major doc bugfix
* show-doc would fail on some core classes, i.e `show-doc Bignum`. This is now fixed
and show allow a wider range of core documentation to be viewed directly in Pry.

18/4/2012 version 0.9.9.2 minor bugfix
* make correct_indentation's auto-colorization respect Pry.color

18/4/2012 version 0.9.9.1 very minor release
* cleared up confusion in show-source/show-doc docs that -a switch applies to classes
as well as modules

18/4/2012 version 0.9.9
MAJOR NEW FEATURES
* lines of input are syntax highlighted upon 'enter' keypress
* show-source command can now show class/module sourcecode (use -a to see all monkeypatches). Hard dependency on ruby18_source_location gem in MRI 1.8
* show-doc command can show class/module docs (use -a to see docs for all monkeypatches) Hard dependency on ruby18_source_location gem in MRI 1.8.
* new `find-method` command, performs a recursive search in a namespace for the existence of methods. Can find methods whose names match a regex or methods which contain provided CODE! This command is like a ruby-aware 'grep', very cool (thanks swarley)
* pry-coolline now works properly with Pry (https://github.com/pry/pry-coolline)
* alias_command method now much more powerful, e.g: alias_command "lM", "ls -M"
* `whereami` is now more intelligent, automatically showing entire sourcecode of current method if current context is a method (thanks robgleeson).
* new `raise-up` command. Allows you to raise an exception that will bubble out of pry (ending the session) and escape into enclosing program.

remaining items:
* fixed windows crashing bug when paging
* lines ending with \ are incomplete (kudos to fowl)
* `edit-method -n` no longer blocks (thanks misfo)s
* show instance methods of modules by default in ls
* docs for REPL defined methods can now be displayed using show-doc
* autoload ruby18_source_location on mri18, when available (https://github.com/conradirwin/ruby18_source_location)
* tab completion should work on first line now (historic bug fixed)
* :quiet => true option added to `Pry.start`, turns off whereami
* another easter egg added
* show unloaded constants in yellow for ls
* improved documentation for Pry.config options
* improved auto indentation
* JRuby: heuristics used to clean up 'ls' output (less internal methods polluting output)

6/3/2012 version 0.9.8.4 major bugfix
* ~/.pry_history wasnt being created (if it did not exist)! FIXED
* `hist --save` saved colors! FIXED
* added Pry#add_sticky_local API for adding sticky locals to individual pry instances

2/3/2012 version 0.9.8.3 minor update
* various tweaks to improve rbx support
* commands now support optional block arguments
* much improved help command
* updated method_source dependencya
* added wtf command
* jruby should now work in windows (though without color)

9/2/2012 version 0.9.8.2 bugfix
* fixed bugs related to --super
* upgraded slop dependency
* added edit -c (edit current line)
* edit now respects Pry.config.disable_autoreload option

30/1/2012 version 0.9.8.1 bugfix
* fixed broken --no-plugins option
* Ensure ARGV is not mutated during option parsing.
* Use a more rbx-friendly test for unicodeness
* Use rbx-{18,19}mode as indicated  http://about.travis-ci.org/docs/user/languages/ruby/
* Don't explode in gem-list [Fixes #453, #454]
* Check for command-name collision on assignment [Fixes #450]

25/1/2012 version 0.9.8

MAJOR NEW FEATURES
- upgraded command api, https://github.com/pry/pry/wiki/Custom-commands
- added a system of hooks for customizing pry behaviour
- changed syntax checking to use eval() for improved accuracy
- added save-file command
- added gist command (removed gist-method, new gist command is more general)

complete CHANGELOG:
* CommandError's no longer cause the current input to be disgarded
* Better syntax highlighting for rbx code code

lib/pry.rb
# (C) John Mair (banisterfiend) 2011
# MIT License
#

require 'pp'
require 'pry/helpers/base_helpers'
require 'pry/hooks'

class Pry
  # The default hooks - display messages when beginning and ending Pry sessions.
  DEFAULT_HOOKS = Pry::Hooks.new.add_hook(:before_session, :default) do |out, target, _pry_|
    next if _pry_.quiet?
    _pry_.run_command("whereami --quiet")
  end

  # The default print
  DEFAULT_PRINT = proc do |output, value|
    output_with_default_format(output, value, :hashrocket => true)
  end

  def self.output_with_default_format(output, value, options = {})
    stringified = begin
                    value.pretty_inspect
                  rescue RescuableException
                    nil
                  end

    unless String === stringified
      # Read the class name off of the singleton class to provide a default
      # inspect.
      klass = (class << value; self; end).ancestors.first
      stringified = "#<#{klass}:0x#{value.__id__.to_s(16)}>"
    end

    nonce = rand(0x100000000).to_s(16) # whatever

    stringified.gsub!(/#</, "%<#{nonce}")
    # Don't recolorize output with color (for cucumber, looksee, etc.) [Issue #751]
    colorized = if stringified =~ /\e\[/
                  stringified
                else
                  Helpers::BaseHelpers.colorize_code(stringified)
                end

    # avoid colour-leak from CodeRay and any of the users' previous output
    colorized = colorized.sub(/(\n*)\z/, "\e[0m\\1") if Pry.color

    result = colorized.gsub(/%<(.*?)#{nonce}/, '#<\1')
    Helpers::BaseHelpers.stagger_output(result, output)
  end

  # may be convenient when working with enormous objects and
  # pretty_print is too slow
  SIMPLE_PRINT = proc do |output, value|
    begin
      output.puts value.inspect
    rescue RescuableException
      output.puts "unknown"
    end
  end

  # useful when playing with truly enormous objects
  CLIPPED_PRINT = proc do |output, value|
    output.puts Pry.view_clip(value)
  end

  # Will only show the first line of the backtrace
  DEFAULT_EXCEPTION_HANDLER = proc do |output, exception, _|
    output.puts "#{exception.class}: #{exception.message}"
    output.puts "from #{exception.backtrace.first}"
  end

  DEFAULT_PROMPT_NAME = 'pry'

  # The default prompt; includes the target and nesting level
  DEFAULT_PROMPT = [
                    proc { |target_self, nest_level, pry|
                      "[#{pry.input_array.size}] #{Pry.config.prompt_name}(#{Pry.view_clip(target_self)})#{":#{nest_level}" unless nest_level.zero?}> "
                    },

                    proc { |target_self, nest_level, pry|
                      "[#{pry.input_array.size}] #{Pry.config.prompt_name}(#{Pry.view_clip(target_self)})#{":#{nest_level}" unless nest_level.zero?}* "
                    }
                   ]

  # A simple prompt - doesn't display target or nesting level
  SIMPLE_PROMPT = [proc { ">> " }, proc { " | " }]

  SHELL_PROMPT = [
                  proc { |target_self, _, _| "#{Pry.config.prompt_name} #{Pry.view_clip(target_self)}:#{Dir.pwd} $ " },
                  proc { |target_self, _, _| "#{Pry.config.prompt_name} #{Pry.view_clip(target_self)}:#{Dir.pwd} * " }
                 ]

  # A prompt that includes the full object path as well as
  # input/output (_in_ and _out_) information. Good for navigation.
  NAV_PROMPT = [
                proc do |conf|
                  tree = conf.binding_stack.map { |b| Pry.view_clip(b.eval("self")) }.join " / "
                  "[#{conf.expr_number}] (#{Pry.config.prompt_name}) #{tree}: #{conf.nesting_level}> "
                end,
                proc do |conf|
                  tree = conf.binding_stack.map { |b| Pry.view_clip(b.eval("self")) }.join " / "
                  "[#{conf.expr_number}] (#{ Pry.config.prompt_name}) #{tree}: #{conf.nesting_level}* "
                end,
               ]

  # Deal with the ^D key being pressed. Different behaviour in different cases:
  #   1. In an expression behave like `!` command.
  #   2. At top-level session behave like `exit` command.
  #   3. In a nested session behave like `cd ..`.
  DEFAULT_CONTROL_D_HANDLER = proc do |eval_string, _pry_|
    if !eval_string.empty?
      eval_string.replace('') # Clear input buffer.
    elsif _pry_.binding_stack.one?
      _pry_.binding_stack.clear
      throw(:breakout)
    else
      # Otherwise, saves current binding stack as old stack and pops last
      # binding out of binding stack (the old stack still has that binding).
      _pry_.command_state["cd"] ||= OpenStruct.new # FIXME
      _pry_.command_state['cd'].old_stack = _pry_.binding_stack.dup
      _pry_.binding_stack.pop
    end
  end

  DEFAULT_SYSTEM = proc do |output, cmd, _|
    if !system(cmd)
      output.puts "Error: there was a problem executing system command: #{cmd}"
    end
  end

  # Store the current working directory. This allows show-source etc. to work if
  # your process has changed directory since boot. [Issue #675]
  INITIAL_PWD = Dir.pwd

  # As a REPL, we often want to catch any unexpected exceptions that may have
  # been raised; however we don't want to go overboard and prevent the user
  # from exiting Pry when they want to.
  module RescuableException
    def self.===(exception)
      case exception
        # Catch when the user hits ^C (Interrupt < SignalException), and assume
        # that they just wanted to stop the in-progress command (just like bash etc.)
      when Interrupt
        true
        # Don't catch signals (particularly not SIGTERM) as these are unlikely to be
        # intended for pry itself. We should also make sure that Kernel#exit works.
      when *Pry.config.exception_whitelist
        false
        # All other exceptions will be caught.
      else
        true
      end
    end
  end

  # Catches SecurityErrors if $SAFE is set
  module TooSafeException
    def self.===(exception)
      $SAFE > 0 && SecurityError === exception
    end
  end

  # Don't catch these exceptions
  DEFAULT_EXCEPTION_WHITELIST = [SystemExit, SignalException, Pry::TooSafeException]

  # CommandErrors are caught by the REPL loop and displayed to the user. They
  # indicate an exceptional condition that's fatal to the current command.
  class CommandError < StandardError; end
  class MethodNotFound < CommandError; end

  # indicates obsolete API
  class ObsoleteError < StandardError; end

  # This is to keep from breaking under Rails 3.2 for people who are doing that
  # IRB = Pry thing.
  module ExtendCommandBundle
  end
end

if Pry::Helpers::BaseHelpers.mri_18?
  begin
  require 'ruby18_source_location'
  rescue LoadError
  end
end

require 'method_source'
require 'shellwords'
require 'stringio'
require 'coderay'
require 'slop'
require 'rbconfig'
require 'tempfile'

begin
  require 'readline'
rescue LoadError
  warn "You're running a version of ruby with no Readline support"
  warn "Please `gem install rb-readline` or recompile ruby --with-readline."
  exit!
end

if Pry::Helpers::BaseHelpers.jruby?
  begin
    require 'ffi'
  rescue LoadError
    warn "Need to `gem install ffi`"
  end
end

if Pry::Helpers::BaseHelpers.windows? && !Pry::Helpers::BaseHelpers.windows_ansi?
  begin
    require 'win32console'
  # The mswin and mingw versions of pry require win32console, so this should
  # only fail on jruby (where win32console doesn't work).
  # Instead we'll recommend ansicon, which does.
  rescue LoadError
    warn "For a better pry experience, please use ansicon: http://adoxa.3eeweb.com/ansicon/"
  end
end

begin
  require 'bond'
rescue LoadError
end

require 'pry/version'
require 'pry/repl'
require 'pry/rbx_method'
require 'pry/rbx_path'
require 'pry/code'
require 'pry/history_array'
require 'pry/helpers'
require 'pry/code_object'
require 'pry/method'
require 'pry/wrapped_module'
require 'pry/history'
require 'pry/command'
require 'pry/command_set'
require 'pry/commands'
require 'pry/custom_completions'
require 'pry/completion'
require 'pry/plugins'
require 'pry/core_extensions'
require 'pry/pry_class'
require 'pry/pry_instance'
require 'pry/cli'
require 'pry/pager'
require 'pry/terminal_info'
require 'pry/editor'
